* add adb and adb_devices action

hjanuschka opened the pull request:
this adds the following actions
- `adb`
  - allows to run a any adb command, specified via `flags:`
- `adb_devices`
  - returns an array of connected android devices/emulators
- `gradle``
  - adds serial option parameter to specify wich device to run on

i needed the actions to port some of my android-CI scripts to fastlane :boom: 

for example to get the model name of all connected devices  - and run androidConnectedTest on all devices.

``` ruby
lane :android_demo do
        adb({adb_path: "/Users/hja/Library/Android/sdk/platform-tools/adb", flags: "kill-server"})
        adb({adb_path: "/Users/hja/Library/Android/sdk/platform-tools/adb", flags: "start-server"})
        devs = adb_devices({adb_path: "/Users/hja/Library/Android/sdk/platform-tools/adb"});

        devs.each do | device |
                model = adb({adb_path: "/Users/hja/Library/Android/sdk/platform-tools/adb", flags: "shell getprop ro.product.model", serial: device.serial})
                puts "Device - #{device.serial} - Model: #{model}"
                gradle({
                    task: "connectedDebugAndroidTest",
                    serial: device.serial
                })
        end
        gradle({task: "no-serial-task"});
end
```

i tried to stick to the coding conventions, but please be nice to me  - it is my first ruby expierence :o !!

truncated sample output:

``` bash
[21:10:29]: Successfully loaded custom action '/Users/hja/fastlane/fastlane/actions/rubocop.rb'.
[21:10:29]: Driving the lane 'android_demo' 🚀
[21:10:29]: -----------------
[21:10:29]: --- Step: adb ---
[21:10:29]: -----------------
[21:10:29]: [SHELL COMMAND]:  /Users/hja/Library/Android/sdk/platform-tools/adb kill-server
[21:10:29]: -----------------
[21:10:29]: --- Step: adb ---
[21:10:29]: -----------------
[21:10:29]: [SHELL COMMAND]:  /Users/hja/Library/Android/sdk/platform-tools/adb start-server
[21:10:32]: -------------------------
[21:10:32]: --- Step: adb_devices ---
[21:10:32]: -------------------------
[21:10:32]: -----------------
[21:10:32]: --- Step: adb ---
[21:10:32]: -----------------
[21:10:32]: [SHELL COMMAND]: ANDROID_SERIAL=emulator-5554 /Users/hja/Library/Android/sdk/platform-tools/adb shell getprop ro.product.model
[21:10:32]: [SHELL]: Android SDK built for x86
[21:10:32]: Device - emulator-5554 - Model: Android SDK built for x86
[21:10:32]: --------------------
[21:10:32]: --- Step: gradle ---
[21:10:32]: --------------------
[21:10:54]: -----------------
[21:10:54]: --- Step: adb ---
[21:10:54]: -----------------
[21:10:54]: [SHELL COMMAND]: ANDROID_SERIAL=SH35CW906378 /Users/hja/Library/Android/sdk/platform-tools/adb shell getprop ro.product.model
[21:10:54]: [SHELL]: One
[21:10:54]: Device - SH35CW906378 - Model: One
[21:10:54]: --------------------
[21:10:54]: --- Step: gradle ---
[21:10:54]: --------------------
[21:10:58]: [SHELL COMMAND]: ANDROID_SERIAL=SH35CW906378 /Users/hja/Desktop/android-project/gradlew tasks 
[21:11:01]: --------------------
[21:11:01]: --- Step: gradle ---
[21:11:01]: --------------------
[21:11:05]: [SHELL COMMAND]:  /Users/hja/Desktop/android-project/gradlew tasks 

+------+-----------------+-------------+
|           fastlane summary           |
+------+-----------------+-------------+
| Step | Action          | Time (in s) |
+------+-----------------+-------------+
| 1    | update_fastlane | 0           |
| 2    | adb             | 0           |
| 3    | adb             | 3           |
| 4    | adb_devices     | 0           |
| 5    | adb             | 0           |
| 6    | gradle          | 21          |
| 7    | adb             | 0           |
| 8    | gradle          | 6           |
| 9    | gradle          | 6           |
+------+-----------------+-------------+

[21:11:07]: fastlane.tools finished successfully 🎉
```

hjanuschka committed:
* add adb and adb_devices action
* add support for serial param in gradle action to run on different devices

hjanuschka commented:
i think the failing travis has nothing todo with this pr

hjanuschka commented:
i am also having a branch wich includes the great pr here #1082  ->  https://github.com/hjanuschka/fastlane/tree/hj_with_new_gradle

hjanuschka commented:
my final lanes to do ui-unit tests on android using fastlane (and robotium as a testing framework)

 * gets the connected devices
 * foreach device it unlocks the device and  runs the `connectedDebugAndroidTest` gradle task
 * creates a device-model folder in reports, mv's the device results
 * grabs the device screenshots
 * locks device


```ruby
private_lane :android_test_on_single_dev do | options |


        adb({ flags: "shell input keyevent 82", serial: options[:serial]})
        #uninstall existing app
        adb({ flags: "shell pm uninstall -k krone.at.fxf", serial: options[:serial]})

        #clear logcat
        adb({ flags: "logcat -c", serial: options[:serial]})
        # run tests
        gradle({
          task: "connectedDebugAndroidTest",
          serial: options[:serial]
          })
        
        
        
        # grab robotium screenshots
        adb({ flags: "pull /storage/sdcard0/Robotium-Screenshots/ app/build/reports/androidTests/connected/screenshots", serial: options[:serial]})
        # on some devices the sdcard is somwhere else
        adb({ flags: "pull /sdcard/Robotium-Screenshots/ app/build/reports/androidTests/connected/screenshots", serial: options[:serial]})
        
        # export logcat to textfile
        adb({ flags: "logcat -d > app/build/reports/androidTests/connected/logcat.txt", serial: options[:serial]})
        
        #turn of screen 
        
        adb({ flags: "shell input keyevent 26", serial: options[:serial]})
        

end
private_lane :android_restart_adb do
  adb({ flags: "kill-server"})
  adb({ flags: "start-server"})
end

lane :android_test do
        require 'fileutils'
        android_restart_adb      
        devs = adb_devices;
        if devs.empty? 
          raise "No Device Connected"
        end
        devs.each do | device |
                model = adb({ flags: "shell getprop ro.product.model", serial: device.serial}).strip
                puts "Device - #{device.serial} - Model: #{model}"
                android_test_on_single_dev({serial: device.serial})

                FileUtils.mkdir_p("../app/build/reports/#{model}")
                FileUtils.cp_r("../app/build/reports/androidTests", "../app/build/reports/#{model}")
                FileUtils.rm_rf("../app/build/reports/androidTests")

        end
        
end
```

asfalcone commented on lib/fastlane/actions/adb.rb:
```diff
@@ -0,0 +1,65 @@
+                                       description: "Android serial, wich device should be used for this command",
+                                       is_string: true,
+                                       default_value: ""),
+          FastlaneCore::ConfigItem.new(key: :flags,
```
Can we call this `command` instead of `flags`? The values passed to `adb` are [documented](http://developer.android.com/tools/help/adb.html) as commands.

asfalcone commented on spec/actions_md_spec.rb:
```diff
@@ -3,7 +3,7 @@
     local_actions = ["rubocop"]
     discontinued = ["update_project_code_signing"]
-    others = ["default_platform", "gradle"]
+    others = ["default_platform", "gradle", "adb", "adb_devices"]
```
Please remove `adb` and `adb_devices` from the exceptions list and add documentation for them to [Actions.md](https://github.com/fastlane/fastlane/blob/master/docs/Actions.md). (Gradle should also be documented but haven't gotten a chance to do that! 🙊)

asfalcone commented on lib/fastlane/actions/adb.rb:
```diff
@@ -0,0 +1,65 @@
+        [
+          FastlaneCore::ConfigItem.new(key: :serial,
+                                       env_name: "FL_ANDROID_SERIAL",
+                                       description: "Android serial, wich device should be used for this command",
```
Typo: "wich" should be "which"

asfalcone commented:
@hjanuschka Thanks for this PR and sorry for the delay in reviewing! I've tested it out and it works great. I've also added a few comments which need to be addressed before it can be merged.

hjanuschka commented:
@asfalcone thx for taking the time to review this.

  * rebased current master
  * done the changes from your comments
  * added the docs/Actions.md entry
  * no hound woof
  * tests should pass

hopefully this is mergeable now.

asfalcone commented:
@hjanuschka :tada:

hjanuschka commented:
so nice thx @asfalcone
